### PR TITLE
Mean sd

### PR DESCRIPTION
**Pull request by mattansb**
I always write a function for computing the mean+-sd, and I just WANT IT TO EXIST SOMEWHERE!

What do you think?

``` r
library(datawizard)

mean_sd(mtcars$mpg)
#>      -SD     Mean      +SD 
#> 14.06368 20.09062 26.11757

median_mad(mtcars$mpg)
#>     -MAD   Median     +MAD 
#> 13.78851 19.20000 24.61149
```

<sup>Created on 2023-01-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

File: R/mean_sd.R [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#'
+#' median_mad(mtcars$mpg)
+#'
+#' @export
+mean_sd <- function(x, na.rm = TRUE, named = TRUE, ...) {
+  x <- as.numeric(x)
+
+  M <- mean(x, na.rm = na.rm)
+  S <- sd(x, na.rm = na.rm)
+
```
Comment (strengejacke) [inline R/mean_sd.R:22]: `stats::sd`

File: R/mean_sd.R [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+#' @export
+#' @rdname mean_sd
+median_mad <- function(x, na.rm = TRUE, constant = 1.4826, named = TRUE, ...) {
+  x <- as.numeric(x)
+
+  M <- median(x, na.rm = na.rm)
+  S <- mad(x, na.rm = na.rm, constant = constant)
+
```
Comment (strengejacke) [inline R/mean_sd.R:37]: `stats::median`, `stats::mad`.

**Comment by strengejacke** [comment]
``` r
centrality_dispersion <- function(x,
                                  type = "mean",
                                  na.rm = TRUE,
                                  times = 1,
                                  constant = 1.4,
                                  named = TRUE,
                                  ...) {
  x <- as.numeric(x)
  type <- match.arg(type, choices = c("mean", "median"))

  # centrality
  M <- switch(type,
    median = stats::median(x, na.rm = na.rm),
    mean(x, na.rm = na.rm)
  )

  S <- switch(type,
    median = stats::mad(x, na.rm = na.rm, constant = constant),
    stats::sd(x, na.rm = na.rm)
  )
  
  v <- M + c(-rev(seq_len(times)), 0, seq_len(times)) * S
  
  if (isTRUE(named)) {
     string_cs <- switch(type,
      median = c("Median", "MAD"),
      c("Mean", "SD")
    )
    names(v) <- c(
      paste0("-", rev(seq_len(times)), " ", string_cs[2]),
      string_cs[1],
      paste0("+", seq_len(times), " ", string_cs[2])
    )
  }
  v
}

mean_sd <- function(x, na.rm = TRUE, times = 1, named = TRUE, ...) {
  centrality_dispersion(x, type = "mean", na.rm = na.rm, times = times, named = named, ...)
}

median_mad <- function(x, na.rm = TRUE, times = 1, constant = 1.4, named = TRUE, ...) {
  centrality_dispersion(x, type = "median", na.rm = na.rm, times = times, constant = constant, named = named, ...)
}

centrality_dispersion(mtcars$mpg)
#>    -1 SD     Mean    +1 SD 
#> 14.06368 20.09062 26.11757
centrality_dispersion(mtcars$mpg, times = 3)
#>     -3 SD     -2 SD     -1 SD      Mean     +1 SD     +2 SD     +3 SD 
#>  2.009781  8.036729 14.063677 20.090625 26.117573 32.144521 38.171469
centrality_dispersion(mtcars$mpg, type = "median", times = 2)
#> -2 MAD -1 MAD Median +1 MAD +2 MAD 
#>   8.98  14.09  19.20  24.31  29.42

mean_sd(mtcars$mpg)
#>    -1 SD     Mean    +1 SD 
#> 14.06368 20.09062 26.11757

median_mad(mtcars$mpg, times = 2)
#> -2 MAD -1 MAD Median +1 MAD +2 MAD 
#>   8.98  14.09  19.20  24.31  29.42
```

<sup>Created on 2023-01-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by mattansb** [comment]
I don't see a need for a "generic" - calling it would be more complicated than calling the right function.

Also, have you ever needed `times > 1`?

**Comment by strengejacke** [comment]
> I don't see a need for a "generic" - calling it would be more complicated than calling the right function.

We can make it internal, so we avoid repeated code.

> 
> Also, have you ever needed `times > 1`?

Maybe not me, but I bet you'll find a user who needs it 😉

**Comment by mattansb** [comment]
This can be merged - all failing tests are unrelated to this PR.

**Review by etiennebacher — CHANGES_REQUESTED** [review]
Thanks, can you also add tests for these two functions?

File: NEWS.md [hunk 3]
```diff
@@ -1,3 +1,11 @@
+# datawizard 0.6.5.xx
```
Comment (etiennebacher) [inline NEWS.md:1]: can you also bump the version in DESCRIPTION?

File: R/mean_sd.R [hunk 4]
```diff
@@ -0,0 +1,74 @@
+#' Summary Helpers
+#'
+#' @param x A numeric vector (or one that can be coerced to one via
+#'   `as.numveric()`) to be summarized.
+#' @param named Should the vector be named?
+#'   (E.g., `c("-SD" = -1, Mean = 1, "+SD" = 2)`.)
+#' @param times How many SDs above and below the Mean (or MADs around the Median)
+#' @param ... Not used.
```
[the remark this answers is not in the document]
Comment (mattansb) [reply R/mean_sd.R:8]: I find it to be a general good practice, but we can drop it if you want.

File: tests/testthat/test-mean_sd.R [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# library(testthat)
+
+test_that("mean_sd", {
+  x <- c(-1, 0, 1)
+  msd <- mean_sd(x)
+  expect_equal(unname(msd), x)
+  expect_equal(names(msd), c("-SD", "Mean", "+SD"))
```
Comment (etiennebacher) [inline tests/testthat/test-mean_sd.R:7]: you can use `expect_named()` instead (which is better according to `lintr`)

File: R/mean_sd.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (strengejacke) [reply R/mean_sd.R:8]: You could have a use for `...` when passing arguments down to other functions, even if your function is no generic.

**Review by etiennebacher — APPROVED** [review]

**Comment by etiennebacher** [comment]
Thanks @mattansb!

**Pull request merged**